Hi @mgrbyte there is a bit of a problem. Probably because you test things in the repl rather than starting the server from scratch.

```
Exception in thread "main" java.lang.RuntimeException: Unable to resolve symbol: db in this context, compiling:(wormbase/names/recent.clj:135:14)
	at clojure.lang.Compiler.analyze(Compiler.java:6792)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3881)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7005)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$IfExpr$Parser.parse(Compiler.java:2822)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.access$300(Compiler.java:38)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6368)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6420)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6420)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6420)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6420)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$FnMethod.parse(Compiler.java:5460)
	at clojure.lang.Compiler$FnExpr.parse(Compiler.java:4022)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7001)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3881)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7005)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$MapExpr.parse(Compiler.java:3097)
	at clojure.lang.Compiler.analyze(Compiler.java:6781)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3881)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7005)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$VectorExpr.parse(Compiler.java:3253)
	at clojure.lang.Compiler.analyze(Compiler.java:6775)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6420)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6420)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6420)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6420)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$FnMethod.parse(Compiler.java:5460)
	at clojure.lang.Compiler$FnExpr.parse(Compiler.java:4022)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7001)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3813)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7005)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3881)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7005)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:6100)
	at clojure.lang.Compiler$FnMethod.parse(Compiler.java:5460)
	at clojure.lang.Compiler$FnExpr.parse(Compiler.java:4022)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7001)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$NewExpr$Parser.parse(Compiler.java:2659)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$MapExpr.parse(Compiler.java:3097)
	at clojure.lang.Compiler.analyze(Compiler.java:6781)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3881)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7005)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6991)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3881)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7005)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.access$300(Compiler.java:38)
	at clojure.lang.Compiler$DefExpr$Parser.parse(Compiler.java:595)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:7003)
	at clojure.lang.Compiler.analyze(Compiler.java:6773)
	at clojure.lang.Compiler.analyze(Compiler.java:6729)
	at clojure.lang.Compiler.eval(Compiler.java:7066)
	at clojure.lang.Compiler.load(Compiler.java:7514)
	at clojure.lang.RT.loadResourceScript(RT.java:379)
	at clojure.lang.RT.loadResourceScript(RT.java:370)
	at clojure.lang.RT.load(RT.java:460)
	at clojure.lang.RT.load(RT.java:426)
	at clojure.core$load$fn__6548.invoke(core.clj:6046)
	at clojure.core$load.invokeStatic(core.clj:6045)
	at clojure.core$load.doInvoke(core.clj:6029)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5848)
	at clojure.core$load_one.invoke(core.clj:5843)
	at clojure.core$load_lib$fn__6493.invoke(core.clj:5888)
	at clojure.core$load_lib.invokeStatic(core.clj:5887)
	at clojure.core$load_lib.doInvoke(core.clj:5868)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:659)
	at clojure.core$load_libs.invokeStatic(core.clj:5925)
	at clojure.core$load_libs.doInvoke(core.clj:5909)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:659)
	at clojure.core$require.invokeStatic(core.clj:5947)
	at clojure.core$require.doInvoke(core.clj:5947)
	at clojure.lang.RestFn.invoke(RestFn.java:3894)
	at wormbase.names.service$eval14503$loading__6434__auto____14504.invoke(service.clj:1)
	at wormbase.names.service$eval14503.invokeStatic(service.clj:1)
	at wormbase.names.service$eval14503.invoke(service.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7062)
	at clojure.lang.Compiler.eval(Compiler.java:7051)
	at clojure.lang.Compiler.load(Compiler.java:7514)
	at clojure.lang.RT.loadResourceScript(RT.java:379)
	at clojure.lang.RT.loadResourceScript(RT.java:370)
	at clojure.lang.RT.load(RT.java:460)
	at clojure.lang.RT.load(RT.java:426)
	at clojure.core$load$fn__6548.invoke(core.clj:6046)
	at clojure.core$load.invokeStatic(core.clj:6045)
	at clojure.core$load.doInvoke(core.clj:6029)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5848)
	at clojure.core$load_one.invoke(core.clj:5843)
	at clojure.core$load_lib$fn__6493.invoke(core.clj:5888)
	at clojure.core$load_lib.invokeStatic(core.clj:5887)
	at clojure.core$load_lib.doInvoke(core.clj:5868)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:659)
	at clojure.core$load_libs.invokeStatic(core.clj:5925)
	at clojure.core$load_libs.doInvoke(core.clj:5909)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:659)
	at clojure.core$require.invokeStatic(core.clj:5947)
	at clojure.core$require.doInvoke(core.clj:5947)
	at clojure.lang.RestFn.invoke(RestFn.java:3894)
	at user$eval15.invokeStatic(form-init3549889180002435738.clj:1)
	at user$eval15.invoke(form-init3549889180002435738.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7062)
	at clojure.lang.Compiler.eval(Compiler.java:7051)
	at clojure.lang.Compiler.load(Compiler.java:7514)
	at clojure.lang.Compiler.loadFile(Compiler.java:7452)
	at clojure.main$load_script.invokeStatic(main.clj:278)
	at clojure.main$init_opt.invokeStatic(main.clj:280)
	at clojure.main$init_opt.invoke(main.clj:280)
	at clojure.main$initialize.invokeStatic(main.clj:311)
	at clojure.main$null_opt.invokeStatic(main.clj:345)
	at clojure.main$null_opt.invoke(main.clj:342)
	at clojure.main$main.invokeStatic(main.clj:424)
	at clojure.main$main.doInvoke(main.clj:387)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:702)
	at clojure.main.main(main.java:37)
Caused by: java.lang.RuntimeException: Unable to resolve symbol: db in this context
	at clojure.lang.Util.runtimeException(Util.java:221)
	at clojure.lang.Compiler.resolveIn(Compiler.java:7299)
	at clojure.lang.Compiler.resolve(Compiler.java:7243)
	at clojure.lang.Compiler.analyzeSymbol(Compiler.java:7204)
	at clojure.lang.Compiler.analyze(Compiler.java:6752)
	... 235 more
Subprocess failed
```

--- a/src/wormbase/names/auth.clj
+++ b/src/wormbase/names/auth.clj
@@ -95,9 +95,7 @@
         parsed-token (parse-token token)
         db (:db request)]
     (if-let [person (verified-person db auth-token-conf parsed-token)]
-      (do
-        (println "Found stored token")
-        (Identification. parsed-token person))
+      (Identification. parsed-token person)
       (when-let [tok (verify-token token)]
         (do
           (println "Checking for person in db with email:" (:email tok))

--- a/src/wormbase/names/auth.clj
+++ b/src/wormbase/names/auth.clj
@@ -95,16 +95,20 @@
         parsed-token (parse-token token)
         db (:db request)]
     (if-let [person (verified-person db auth-token-conf parsed-token)]
-      (Identification. parsed-token person)
+      (do
+        (println "Found stored token")
+        (Identification. parsed-token person))
       (when-let [tok (verify-token token)]
-        (if-let [person (query-person db :person/email (:email tok))]
-          (let [auth-token (sign-token auth-token-conf tok)
-                tx-result @(d/transact-async (:conn request)
-                                             [[:db/add
-                                               (:db/id person)
-                                               :person/auth-token
-                                               auth-token]])]
-            (Identification. parsed-token person)))
+        (do
+          (println "Checking for person in db with email:" (:email tok))
+          (if-let [person (query-person db :person/email (:email tok))]
+            (let [auth-token (sign-token auth-token-conf tok)
+                  tx-result @(d/transact-async (:conn request)
+                                               [[:db/add
+                                                 (:db/id person)
+                                                 :person/auth-token
+                                                 auth-token]])]
+              (Identification. parsed-token person))))
         (log/warn (str "No matching token in db"))))))
 
 (def backend (babt/token-backend {:authfn identify}))

--- a/src/wormbase/names/recent.clj
+++ b/src/wormbase/names/recent.clj
@@ -37,9 +37,9 @@
 (def imported-date (memoize find-max-imported-date))
 
 (defn query-activities
-  ([db log rules ^Date from ^Date until]
-   (query-activities db log rules "" from until))
-  ([db log rules needle ^Date from ^Date until]
+  ([db log rules ^Date from ^Date until how]
+   (query-activities db log rules "" from until how))
+  ([db log rules needle ^Date from ^Date until how]
    ;; find the date for the most recent transaction after imported transactions.
    (let [import-date (imported-date db)
          ;; choose the date that is older betweem thn requested date and last import tx
@@ -52,12 +52,12 @@
          ;; jvm (cold): 107.266427 msecs
          ;; jvm (warm): 30.054339 msecs
          query '[:find ?tx ?e
-                 :in $ % ?log ?since-start ?since-end ?needle
+                 :in $ % ?log ?since-start ?since-end ?needle ?how
                  :where
                  [(tx-ids ?log ?since-start ?since-end) [?tx ...]]
-                 (filter-events ?tx ?needle)
+                 (filter-events ?tx ?needle ?how)
                  [(tx-data ?log ?tx) [[?e]]]]]
-     (d/q query db rules log since-t now needle))))
+     (d/q query db rules log since-t now needle how))))
 
 (defn changes-and-prov-puller
   "Creates a transducer for pulling changes and provenance."
@@ -86,25 +86,24 @@
   "Return recent activities for both batch and individual operations.
   The result should be map whose keys represent these two groupings.
   The groupings in turn should be a sequence of maps."
-  ([db log rules puller ^Date from ^Date until]
-   (activities db log rules puller "" from until))
-  ([db log rules puller needle ^Date from ^Date until]
-   (->> (query-activities db log rules from until)
-        (sequence puller)
-        (remove nil?))))
+  [db log rules puller needle how ^Date from ^Date until]
+  (->> (query-activities db log rules from until how)
+       (sequence puller)
+       (remove nil?)))
 
-(def entity-rules '[[(filter-events ?tx ?needle)
+(def entity-rules '[[(filter-events ?tx ?needle ?how)
                      [(missing? $ ?tx :batch/id)]
                      [?tx :provenance/what ?wid]
+                     [?tx :provenance/how ?how]
                      [?wid :db/ident ?what]
                      [(name ?what) ?w]
                      [(clojure.string/includes? ?w ?needle)]]])
 
-(def person-rules '[[(filter-events ?tx ?needle)
+(def person-rules '[[(filter-events ?tx ?needle _)
                      [?tx :provenance/who ?pid]
                      [?pid :person/email ?needle]]])
 
-(def batch-rules '[[(filter-events ?tx ?needle)
+(def batch-rules '[[(filter-events ?tx ?needle _)
                     [?tx :batch/id _ _ ]]])
 
 
@@ -117,19 +116,21 @@
   (-> etag codecs/str->bytes b64/decode codecs/bytes->str))
 
 (defn handle
-  [request rules puller needle from until]
-  (let [{conn :conn db :db} request
-        log (d/log conn)
-        from* (or from (since-days-ago *default-days-ago*))
-        until* (or until (jt/to-java-date (jt/instant)))
-        items (->> (activities db log rules puller (or needle "") from* until*)
-                   (map (partial wu/elide-db-internals db))
-                   (sort-by :t))
-        latest-t (some-> items first :t)
-        etag (encode-etag latest-t)]
-    (-> {:activities items}
-        (ok)
-        (header "etag" etag))))
+  ([request rules puller needle from until]
+   (handle request rules puller needle #{:agent/console :agent/web} from until))
+  ([request rules puller needle how from until]
+   (let [{conn :conn db :db} request
+         log (d/log conn)
+         from* (or from (since-days-ago *default-days-ago*))
+         until* (or until (jt/to-java-date (jt/instant)))
+         items (->> (activities db log rules puller (or needle "") how from* until*)
+                    (map (partial wu/elide-db-internals db))
+                    (sort-by :t))
+         latest-t (some-> items first :t)
+         etag (encode-etag latest-t)]
+     (-> {:activities items}
+         (ok)
+         (header "etag" etag)))))
 
 (def routes (sweet/routes
              (sweet/context "/recent" []
@@ -159,19 +160,25 @@
                (sweet/GET "/gene" request
                  :tags ["recent" "gene"]
                  :summary "List recent gene activity."
-                 (handle request
-                         entity-rules
-                         (changes-and-prov-puller request)
-                         "gene"
-                         from
-                         until))
+                 :query-params [how :- ::wsr/how]
+                 (do
+                   (println "HOW:" how)
+                   (handle request
+                           entity-rules
+                           (changes-and-prov-puller request)
+                           "gene"
+                           how
+                           from
+                           until)))
                (sweet/GET "/variation" request
                  :tags ["recent" "variation"]
                  :summary "List recent variation activity."
+                 :query-params [how :- ::wsr/how]
                  (handle request
                          entity-rules
                          (changes-and-prov-puller request)
                          "variation"
+                         how
                          from
                          until)))))
 

--- a/src/wormbase/names/recent.clj
+++ b/src/wormbase/names/recent.clj
@@ -148,7 +148,7 @@
                  :query-params [id :- :person/id]
                  :summary "List recent activities made by the currently logged-in user."
                  (when-let [person (if id
-                                     (d/pull db [:person/email] [:person/id id])
+                                     (d/pull (:db request) [:person/email] [:person/id id])
                                      (some-> request :identity :person))]
                    (handle request
                            person-rules

--- a/src/wormbase/names/recent.clj
+++ b/src/wormbase/names/recent.clj
@@ -161,15 +161,13 @@
                  :tags ["recent" "gene"]
                  :summary "List recent gene activity."
                  :query-params [how :- ::wsr/how]
-                 (do
-                   (println "HOW:" how)
-                   (handle request
-                           entity-rules
-                           (changes-and-prov-puller request)
-                           "gene"
-                           how
-                           from
-                           until)))
+                 (handle request
+                         entity-rules
+                         (changes-and-prov-puller request)
+                         "gene"
+                         how
+                         from
+                         until))
                (sweet/GET "/variation" request
                  :tags ["recent" "variation"]
                  :summary "List recent variation activity."

--- a/src/wormbase/specs/recent.clj
+++ b/src/wormbase/specs/recent.clj
@@ -1,11 +1,17 @@
 (ns wormbase.specs.recent
   (:require
    [clojure.spec.alpha :as s]
+   [spec-tools.spec :as sts]
    [spec-tools.core :as stc]
+   [wormbase.specs.agent :as wsa]
    [wormbase.specs.provenance :as wsp]))
 
 (s/def ::from (s/nilable inst?))
 
 (s/def ::until (s/nilable inst?))
 
-(s/def ::activities (stc/spec ::wsp/history))
+(s/def ::activities (stc/spec {:spec (s/coll-of ::wsp/temporal-change)}))
+
+(s/def ::how (stc/spec {:spec (s/and sts/keyword? #{:agent/console :agent/web})
+                        :into #{}
+                        :type :keyword}))

